File: python-prototype/game.py
"""贪吃蛇的核心游戏逻辑。

这个文件只回答一个问题：**下一步，蛇会变成什么样？**

它不打印任何东西，也不知道终端、画布、键盘的存在 —— 那些是 main.py 的事。
把「规则」和「显示」分开写，规则就能脱离界面单独测试，这是本阶段最重要的一课。
"""

import random
from collections import deque
from enum import Enum, auto

# 四个方向，值是从当前格子走到下一个格子的位移 (dx, dy)。
# 注意 y 轴是向下的：up 是 y - 1，不是 y + 1。
DIRECTIONS: dict[str, tuple[int, int]] = {
    "up": (0, -1),
    "down": (0, 1),
    "left": (-1, 0),
    "right": (1, 0),
}

# 每个方向的反方向。蛇不能原地掉头（那等于撞上自己的脖子），
# 所以玩家输入新方向时，要先拿这张表查一遍。
OPPOSITE: dict[str, str] = {
    "up": "down",
    "down": "up",
    "left": "right",
    "right": "left",
}

# 坐标类型别名：一个 (x, y) 的格子
Position = tuple[int, int]

# 吃到一颗食物的得分。写在这里而不是散落在代码里，
# 以后想改成分数随长度递增，只需要动这一处。
SCORE_PER_FOOD = 10


class GameState(Enum):
    """一局游戏当前所处的状态。"""

    READY = auto()      # 还没开始，等玩家按开始
    RUNNING = auto()    # 进行中
    PAUSED = auto()     # 暂停
    GAME_OVER = auto()  # 本局结束


class Death(Enum):
    """蛇的死因。

    这里存英文标记而不是中文，是因为 game.py 管规则、main.py 管显示。
    逻辑里塞中文提示语，将来做网页版就得翻译一遍，早晚会分叉。
    """

    WALL = "wall"  # 撞墙
    SELF = "self"  # 撞到自己


class Snake:
    """一条蛇。

    蛇身用 deque（双端队列）存坐标，**队首是蛇头，队尾是蛇尾**。

    为什么用 deque 而不是 list？因为每走一步都要「头部插入 + 尾部删除」，
    deque 做这两件事都是瞬间完成，而 list 在头部插入需要把后面所有元素
    往后挪一遍。蛇不长的时候感觉不出来，但这属于写之前就该想对的事。
    """

    def __init__(
        self,
        start: Position = (5, 5),
        direction: str = "right",
        length: int = 3,
    ) -> None:
        if direction not in DIRECTIONS:
            raise ValueError(f"未知方向：{direction}")
        if length < 1:
            raise ValueError("蛇至少要有 1 节")

        self.direction = direction
        # 玩家按下的新方向先存在这里，真正移动时才生效。
        # 为什么要两个变量？见 turn() 的注释。
        self.next_direction = direction

        # 身体从蛇头往「身后」依次排开，摆出初始长度。
        #
        # 「身后」就是前进方向的反方向。朝右走身体在左边，朝上走身体在下方。
        # 这里千万别写死成「永远往左」—— 那样朝左走时，头往前一步正好踩在
        # 原来脖子的格子上，一开局就自己撞死自己。
        dx, dy = DIRECTIONS[direction]
        head_x, head_y = start
        self.body: deque[Position] = deque(
            (head_x - dx * i, head_y - dy * i) for i in range(length)
        )

    @property
    def head(self) -> Position:
        """蛇头坐标。"""
        return self.body[0]

    @property
    def tail(self) -> Position:
        """蛇尾坐标。"""
        return self.body[-1]

    def turn(self, direction: str) -> None:
        """记录玩家想转的方向。

        这里有个经典坑：如果不做检查，玩家在一帧之内连按「上」和「左」，
        蛇会先掉头撞上自己。所以和当前方向相反的输入直接忽略。

        注意只改了 next_direction，没有改 direction —— 这样一帧内按五次
        方向键也不会让蛇瞬移，它每帧最多转一次弯。
        """
        if direction not in DIRECTIONS:
            raise ValueError(f"未知方向：{direction}")
        if direction == OPPOSITE[self.direction]:
            return  # 想掉头？当作没听见
        self.next_direction = direction

    def next_head(self) -> Position:
        """按「下一步方向」算出的新蛇头位置 —— 此时蛇还没真的移动。

        为什么要单独一个方法？因为 Game 需要**提前**知道头会走到哪，
        才能判断这一步到底吃没吃到食物。
        """
        dx, dy = DIRECTIONS[self.next_direction]
        return (self.head[0] + dx, self.head[1] + dy)

    def step(self, grow: bool = False) -> Position:
        """往前走一步，返回新的蛇头坐标。

        grow=True 时不砍掉尾巴，蛇就长了一节 —— 吃到食物的时候用。
        这是「吃到食物变长」最简洁的实现方式：变长不是加上一节，
        而是本次不删除尾巴。
        """
        self.direction = self.next_direction
        new_head = self.next_head()

        self.body.appendleft(new_head)  # 新头入队
        if not grow:
            self.body.pop()             # 尾巴出队，长度不变

        return new_head

    def __len__(self) -> int:
        """len(蛇) 得到蛇的长度。"""
        return len(self.body)

    def __repr__(self) -> str:
        return f"Snake(direction={self.direction!r}, body={list(self.body)})"


class Game:
    """一局游戏。

    Snake 只管「蛇自己怎么动」，Game 管它和世界的关系：
    棋盘多大、食物在哪、这一步吃到没有、得了多少分。

    这样分工的好处：以后要加排行榜、加难度等级、加障碍物，
    改动都集中在 Game 里，Snake 不用动。
    """

    def __init__(
        self,
        width: int = 20,
        height: int = 20,
        start: Position = (5, 5),
        direction: str = "right",
        length: int = 3,
        seed: int | None = None,
    ) -> None:
        if width < 2 or height < 2:
            raise ValueError("棋盘至少要有 2x2")

        self.width = width
        self.height = height
        self.score = 0

        # 用自己独立的 Random 实例，而不是 random.choice 这类全局函数。
        # 关键在于：传了 seed 之后，食物出现的顺序每次运行都一模一样。
        # 否则你没法写「第 3 步应该吃到食物」这种测试 —— 随机的东西测不了。
        self.rng = random.Random(seed)

        # 记下开局参数，重开的时候要用
        self.start_position = start
        self.start_direction = direction
        self.start_length = length

        self.snake = Snake(start=start, direction=direction, length=length)
        self.food: Position | None = None
        self.state = GameState.READY
        self.death: Death | None = None
        self.spawn_food()

    @property
    def has_won(self) -> bool:
        """蛇占满整个棋盘就算通关 —— 理论可行，实战几乎遇不到。"""
        return self.food is None

    def free_cells(self) -> list[Position]:
        """棋盘上所有还没被蛇占的格子。"""
        occupied = set(self.snake.body)
        return [
            (x, y)
            for x in range(self.width)
            for y in range(self.height)
            if (x, y) not in occupied
        ]

    def spawn_food(self) -> Position | None:
        """把食物随机放到一个空格子上，返回它的坐标。

        `if not free` 处理的是「蛇把棋盘占满了」的情况。
        这属于理论上可能、实战几乎遇不到的边界情况，但还是要写：
        不写的话，从空列表里随机取一个会直接抛异常，而且报错信息
        完全看不出和「蛇赢了」有什么关系。
        """
        free = self.free_cells()
        if not free:
            self.food = None
            return None
        self.food = self.rng.choice(free)
        return self.food

    def step(self) -> bool:
        """推进一帧，返回这一步是否吃到了食物。

        顺序很讲究，一共四步，顺序不能换：

        1. 只有 RUNNING 才动 —— 暂停中和已结束时调用它，什么都不该发生
        2. 先算出头**将要**走到哪，再比对食物位置（不能先移动后判断：
           移动完蛇头已经站在食物格子上了，你没法区分它是刚到的还是本来就在那）
        3. 判断这一步会不会死，会死就直接结束，不做移动
        4. 都不死，才真的移动、加分、补食物
        """
        if self.state is not GameState.RUNNING:
            return False

        next_head = self.snake.next_head()
        ate = next_head == self.food

        death = self.collision_at(next_head, growing=ate)
        if death is not None:
            self.state = GameState.GAME_OVER
            self.death = death
            return False

        self.snake.step(grow=ate)

        if ate:
            self.score += SCORE_PER_FOOD
            self.spawn_food()

        return ate

    def collision_at(self, position: Position, growing: bool = False) -> Death | None:
        """检查走到 position 这个格子会不会死。安全则返回 None。

        growing=True 表示这一步会吃到食物、尾巴不会让开。
        """
        x, y = position
        if not (0 <= x < self.width and 0 <= y < self.height):
            return Death.WALL

        blocking = set(self.snake.body)
        if not growing:
            # 这一步尾巴会让开身后那一格，所以「头撞到自己的尾巴尖」不算死。
            # 这是贪吃蛇最容易写错的细节之一：差一格就是差一条命。
            blocking.discard(self.snake.tail)

        if position in blocking:
            return Death.SELF

        return None

    def start(self) -> None:
        """开始游戏。只在 READY 状态下有效。"""
        if self.state is GameState.READY:
            self.state = GameState.RUNNING

    def toggle_pause(self) -> None:
        """在运行和暂停之间切换。其他状态下按了没反应。"""
        if self.state is GameState.RUNNING:
            self.state = GameState.PAUSED
        elif self.state is GameState.PAUSED:
            self.state = GameState.RUNNING

    def reset(self) -> None:
        """重开一局：分数清零，蛇回到起点，状态回到 READY。

        注意这里是重新造了一条 Snake，而不是把旧蛇的 body 清空重填。
        新建对象比「手动把每个字段改回初始值」可靠 —— 漏改一个字段，
        就会出现「重开之后蛇还带着上一局身长」这种诡异现象。
        """
        self.score = 0
        self.death = None
        self.snake = Snake(
            start=self.start_position,
            direction=self.start_direction,
            length=self.start_length,
        )
        self.food = None
        self.spawn_food()
        self.state = GameState.READY

    def __repr__(self) -> str:
        return (
            f"Game(state={self.state.name}, score={self.score}, "
            f"length={len(self.snake)}, head={self.snake.head}, "
            f"food={self.food}, death={self.death})"
        )

File: web/public/game.js
/**
 * 贪吃蛇的核心游戏逻辑 —— game.py 的 JavaScript 版本。
 *
 * 这个文件里没有一行代码碰过浏览器：不操作 DOM、不画图、不读键盘。
 * 所以它既能在浏览器里被 main.js 使用，也能被 node --test 直接测试。
 *
 * 对照 python-prototype/game.py 读会有意思：同一套规则，两种语言
 * 各自的写法、各自的坑。
 */

export const DIRECTIONS = {
  up: [0, -1],
  down: [0, 1],
  left: [-1, 0],
  right: [1, 0],
};

const OPPOSITE = { up: "down", down: "up", left: "right", right: "left" };

export const GameState = {
  READY: "READY",
  RUNNING: "RUNNING",
  PAUSED: "PAUSED",
  GAME_OVER: "GAME_OVER",
};

export const Death = { WALL: "wall", SELF: "self" };

export const SCORE_PER_FOOD = 10;

/**
 * 把坐标转成字符串，用来当集合的键。
 *
 * 这里藏着 Python 和 JS 最本质的差异之一：
 * Python 的元组 (1, 2) 可以直接塞进 set，两个值相等就算同一个元素。
 * JS 的数组不行 —— [1, 2] !== [1, 2]，它们是两个不同的对象，
 * 哪怕内容一模一样。所以 JS 里判断「这个格子占没占」，
 * 得先把坐标转成 "1,2" 这种字符串。
 */
export const keyOf = ([x, y]) => `${x},${y}`;

/** 两个坐标是不是同一个格子。注意先判 null —— 食物可能不存在。 */
export const sameCell = (a, b) =>
  a !== null && b !== null && a[0] === b[0] && a[1] === b[1];

/**
 * 一个可以指定种子的小型随机数发生器。
 *
 * JS 自带的 Math.random() 没法指定种子，也就是说它不可复现 ——
 * 测试里没法断言「第 3 步会吃到食物」。Python 的 random.Random(seed)
 * 天生支持这件事，JS 得自己写一个。下面是 mulberry32，十几行就够。
 */
function createRng(seed) {
  if (seed === null || seed === undefined) return Math.random;

  let state = seed >>> 0;
  return () => {
    state = (state + 0x6d2b79f5) | 0;
    let t = Math.imul(state ^ (state >>> 15), 1 | state);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

export class Snake {
  constructor({ start = [5, 5], direction = "right", length = 3 } = {}) {
    if (!Object.hasOwn(DIRECTIONS, direction)) {
      throw new Error(`未知方向：${direction}`);
    }
    if (length < 1) throw new Error("蛇至少要有 1 节");

    this.direction = direction;
    // 玩家按下的新方向先存这里，真正移动时才生效。
    // 两个变量是必须的：只有一个的话，玩家一帧内连按两下方向键，
    // 蛇会瞬间掉头撞死自己。
    this.nextDirection = direction;

    // 身体从蛇头往「身后」排开。身后 = 前进方向的反方向。
    const [dx, dy] = DIRECTIONS[direction];
    const [headX, headY] = start;
    this.body = [];
    for (let i = 0; i < length; i += 1) {
      this.body.push([headX - dx * i, headY - dy * i]);
    }
  }

  get head() {
    return this.body[0];
  }

  get tail() {
    return this.body[this.body.length - 1];
  }

  get length() {
    return this.body.length;
  }

  /** 记录想转的方向。和当前方向相反（掉头）的输入直接忽略。 */
  turn(direction) {
    if (!Object.hasOwn(DIRECTIONS, direction)) {
      throw new Error(`未知方向：${direction}`);
    }
    if (direction === OPPOSITE[this.direction]) return;
    this.nextDirection = direction;
  }

  /** 按下一步方向算出的新蛇头位置 —— 此时蛇还没真的移动。 */
  nextHead() {
    const [dx, dy] = DIRECTIONS[this.nextDirection];
    return [this.head[0] + dx, this.head[1] + dy];
  }

  /** 往前走一步。grow=true 时不砍尾巴，蛇就长了一节。 */
  step(grow = false) {
    this.direction = this.nextDirection;
    const newHead = this.nextHead();

    this.body.unshift(newHead);
    if (!grow) this.body.pop();

    return newHead;
  }
}

export class Game {
  constructor({
    width = 20,
    height = 20,
    start = [5, 5],
    direction = "right",
    length = 3,
    seed = null,
  } = {}) {
    if (width < 2 || height < 2) throw new Error("棋盘至少要有 2x2");

    this.width = width;
    this.height = height;
    this.score = 0;
    this.rng = createRng(seed);

    // 记下开局参数，重开时要用
    this.startPosition = start;
    this.startDirection = direction;
    this.startLength = length;

    this.snake = new Snake({ start, direction, length });
    this.food = null;
    this.state = GameState.READY;
    this.death = null;
    this.spawnFood();
  }

  /** 蛇占满整个棋盘算通关 —— 理论可行，实战几乎遇不到。 */
  get hasWon() {
    return this.food === null;
  }

  /** 棋盘上所有还没被蛇占的格子。 */
  freeCells() {
    const occupied = new Set(this.snake.body.map(keyOf));
    const free = [];
    for (let y = 0; y < this.height; y += 1) {
      for (let x = 0; x < this.width; x += 1) {
        if (!occupied.has(`${x},${y}`)) free.push([x, y]);
      }
    }
    return free;
  }

  /** 把食物随机放到一个空格子上。占满了则返回 null（通关）。 */
  spawnFood() {
    const free = this.freeCells();
    if (free.length === 0) {
      this.food = null;
      return null;
    }
    this.food = free[Math.floor(this.rng() * free.length)];
    return this.food;
  }

  /**
   * 推进一帧，返回这一步是否吃到了食物。
   *
   * 顺序不能换：先算头要去哪 → 判断吃没吃到 → 判断会不会死 → 最后才动。
   * 反过来先移动的话，蛇头已经站在食物格子上了，你就分不清它是刚到的
   * 还是本来就在那。
   */
  step() {
    if (this.state !== GameState.RUNNING) return false;

    const nextHead = this.snake.nextHead();
    const ate = sameCell(nextHead, this.food);

    const death = this.collisionAt(nextHead, ate);
    if (death !== null) {
      this.state = GameState.GAME_OVER;
      this.death = death;
      return false;
    }

    this.snake.step(ate);
    if (ate) {
      this.score += SCORE_PER_FOOD;
      this.spawnFood();
    }

    return ate;
  }

  /**
   * 检查走到 position 会不会死。安全则返回 null。
   *
   * growing=true 表示这一步会吃到食物、尾巴不会让开。
   */
  collisionAt(position, growing = false) {
    const [x, y] = position;
    if (x < 0 || x >= this.width || y < 0 || y >= this.height) {
      return Death.WALL;
    }

    const blocking = new Set(this.snake.body.map(keyOf));
    if (!growing) {
      // 这一步尾巴会让开身后那一格，所以「头撞到自己的尾巴尖」不算死。
      blocking.delete(keyOf(this.snake.tail));
    }

    return blocking.has(keyOf(position)) ? Death.SELF : null;
  }

  start() {
    if (this.state === GameState.READY) this.state = GameState.RUNNING;
  }

  togglePause() {
    if (this.state === GameState.RUNNING) this.state = GameState.PAUSED;
    else if (this.state === GameState.PAUSED) this.state = GameState.RUNNING;
  }

  /** 重开一局。重新造一条蛇，比逐个字段改回去可靠。 */
  reset() {
    this.score = 0;
    this.death = null;
    this.snake = new Snake({
      start: this.startPosition,
      direction: this.startDirection,
      length: this.startLength,
    });
    this.food = null;
    this.spawnFood();
    this.state = GameState.READY;
  }
}
